pytest 9.1.0: new `parametrize` `argvalues` typing produces a spurious mypy error for a list of class objects with differing constructor signatures

pytest 9.1.0 added typed overloads for `pytest.mark.parametrize`, typing `argvalues` as `Collection[ParameterSet | Sequence[object] | object]` (part of the deprecation of non-`Collection` iterables, #13409).

Passing a list of two unrelated class objects whose constructors have incompatible parameter kinds (one keyword-only, one positional-or-keyword) now produces a false-positive mypy error:

```
error: List item 0 has incompatible type "type[Foo]"; expected "type[object]"  [list-item]
```

The same call type-checked fine on pytest 9.0.3.
The usage is exactly the form the changelog recommends: a `list` is a `Collection`, and the deprecation targeted iterators (generators) that get exhausted, not lists of classes.

## Environment

- pytest 9.1.0
- mypy 2.1.0
- Python 3.14

## Minimal reproduction

```python
import pytest


class KwOnly:
    def __init__(self, *, x: int = 0) -> None: ...


class Positional:
    def __init__(self, x: int = 0) -> None: ...


# mypy error: List item 0 has incompatible type "type[KwOnly]"; expected "type[object]"
@pytest.mark.parametrize(argnames="cls", argvalues=[KwOnly, Positional])
def test_list(cls: type) -> None: ...


# OK: a tuple keeps each element's own type, so there is no join.
# Note: ruff's PT007 defaults to expecting a list for argvalues, so ruff users
# would have to set flake8-pytest-style.parametrize-values-type = "tuple".
@pytest.mark.parametrize(argnames="cls", argvalues=(KwOnly, Positional))
def test_tuple(cls: type) -> None: ...
```

Hmm, I am having a hard time understanding mypy's error here. I tried with pyright, ty, pyrefly (`-p strict`) and zuban (`--strict`) and they are fine with it.

If I had to guess mypy's logic:
- The type of `KwOnly` is `type[KwOnly]`
- The type of `Positional` is `type[Positional]`
- Their join is `type[object]`, because `type[]` is covariant ([ref](https://typing.python.org/en/latest/spec/special-types.html#type:~:text=type%20is%20covariant%20in%20its%20parameter))
- So the type of the list is `list[type[object]]`
- But then what's this about "incompatible type "type[KwOnly]"; expected "type[object]""?

I think this might be a mypy bug, since it gives this error even without pytest's involvement:

```py
class KwOnly:
    def __init__(self, *, x: int = 0) -> None: ...

class Positional:
    def __init__(self, x: int = 0) -> None: ...

# error: List item 0 has incompatible type "type[KwOnly]"; expected "type[object]"  [list-item]
[KwOnly, Positional]
```

It's interesting it wasn't a problem for you with previous pytest typing.

I have a similar problem:

```python
import pytest

@pytest.mark.parametrize(
    ('schema_path', 'expected_exception'),
    [
        ('does.not.exist:schema', ImportError),
        ('server.urls:urlpatterns', AttributeError),
    ],
)
def test_export_schema_invalid_input(
    schema_path: str,
    expected_exception: type[Exception],
) -> None:
    ...
```

Error:

```
tests/test_integration/test_management/test_dmr_export_schema.py:99: error: List item 1 has incompatible type "tuple[str, type[AttributeError]]"; expected "tuple[str, type[Exception]]"  [list-item]
```

Not quite sure on how to fix it.

@sobolevn Yes it's the same thing:

```py
# error: List item 0 has incompatible type "type[ImportError]"; expected "type[Exception]"  [list-item]
[ImportError, AttributeError]
```

As @adamtheturtle said, it's related to the type's `__init__` signature, so it doesn't happen with e.g. `[TypeError, AttributeError]`.

It does seem that mypy devolves `type[..]` to the ctor signature, e.g. `reveal_type(KwOnly)` gives `Revealed type is "def (*, x: int =) -> KwOnly"`.

I was curious why it worked before. Seems the difference is the addition of `@overload`:

```py
def ok(x: Collection[object]) -> None: pass

ok([KwOnly, Positional])


@overload
def bad(x: Collection[object]) -> None: pass

@overload
def bad(x: object) -> None: pass

def bad(x: object) -> None: pass

# error: List item 0 has incompatible type "type[KwOnly]"; expected "type[object]"  [list-item]
bad([KwOnly, Positional])
```

I would guess that without overload, mypy can infer the argument in the context of `Collection[object]`, and the context makes it work, but with overloads it needs to infer the type without a context first, which gives the error.

I added the overload so I can stick a `@deprecated` on the non-`Collection` overload. If we want to workaround this mypy bug(?), we can either

1, Drop the `Collection` overload, leaving only the runtime deprecation warning
2. Drop the non-`Collection` overload, turning the deprecated usage to a static type error rather than a static deprecation warning

Strictly speaking 2 is not backward compatible so needs to be 1 until pytest 10. I will try to do it for 9.1.1.

--- a/changelog/14606.bugfix.rst
+++ b/changelog/14606.bugfix.rst
@@ -1,0 +1,1 @@
+Fixed ``list-item`` typing errors from mypy in :ref:`@pytest.mark.parametrize <pytest.mark.parametrize ref>` ``argvalues`` parameter.

--- a/src/_pytest/mark/structures.py
+++ b/src/_pytest/mark/structures.py
@@ -21,7 +21,6 @@ from typing import TypeVar
 import warnings
 
 from .._code import getfslineno
-from ..compat import deprecated
 from ..compat import NOTSET
 from ..compat import NotSetType
 from _pytest.config import Config
@@ -537,31 +536,16 @@ if TYPE_CHECKING:
         ) -> MarkDecorator: ...
 
     class _ParametrizeMarkDecorator(MarkDecorator):
-        @overload  # type: ignore[override,no-overload-impl]
-        def __call__(
-            self,
-            argnames: str | Sequence[str],
-            argvalues: Collection[ParameterSet | Sequence[object] | object],
-            *,
-            indirect: bool | Sequence[str] = ...,
-            ids: Iterable[None | str | float | int | bool | _HiddenParam]
-            | Callable[[Any], object | None]
-            | None = ...,
-            scope: ScopeName | None = ...,
-        ) -> MarkDecorator: ...
-
-        @overload
-        @deprecated(
-            "Passing a non-Collection iterable to the 'argvalues' parameter of @pytest.mark.parametrize is deprecated. "
-            "Convert argvalues to a list or tuple.",
-        )
-        def __call__(
+        def __call__(  # type: ignore[override]
             self,
             argnames: str | Sequence[str],
             argvalues: Iterable[ParameterSet | Sequence[object] | object],
+            # TODO(pytest10): Change to below after PARAMETRIZE_NON_COLLECTION_ITERABLE deprecation.
+            #                 Overload doesn't work, see #14606.
+            # argvalues: Collection[ParameterSet | Sequence[object] | object],
             *,
             indirect: bool | Sequence[str] = ...,
-            ids: Iterable[None | str | float | int | bool]
+            ids: Iterable[None | str | float | int | bool | _HiddenParam]
             | Callable[[Any], object | None]
             | None = ...,
             scope: ScopeName | None = ...,

--- a/testing/typing_checks.py
+++ b/testing/typing_checks.py
@@ -67,15 +67,14 @@ def test_hidden_param(x: int) -> None:
     pass
 
 
-# Test @pytest.mark.parametrize iterator argvalues deprecation.
-# Will be complain about unused type ignore if doesn't work.
-@pytest.mark.parametrize("x", iter(range(10)))  # type: ignore[deprecated]
-def test_it(x: int) -> None:
-    pass
-
-
 # Issue #14137.
 def check_scope_typing() -> None:
 
     custom_scope: ScopeName = "function"
     assert_type(custom_scope, ScopeName)
+
+
+# Issue #14606.
+@pytest.mark.parametrize("x", [ImportError, AttributeError])
+def check_mypy_bug_with_argvalues(x) -> None:
+    pass

